docs: generate API reference doc from an OpenAPI spec (#532)
canonical/pebble: docs/reference/api.md

# API

Pebble exposes an HTTP API that remote clients can use to interact with the daemon. The API has endpoints for starting and stopping services, adding configuration layers to the plan, and so on.

The API uses HTTP over a Unix socket, with access to the API controlled by user ID. If `pebble run` is started with the `--http <address>` option, Pebble exposes a limited set of open-access endpoints (see {ref}`api-access-levels`) using the given TCP address.

## Structure of the API

Here is a breakdown of the general structure of most API endpoints in Pebble.

### API endpoint URL

> Pebble API uses standard HTTP URLs to get actions and resources. The URLs are made up of a base URL followed by a path and optional query parameters.

Reviewer: I think we can standardise on "The API" throughout the page, instead of "Pebble API"

Reviewer 2: I think David meant use just "The API", whereas I believe you've changed it to "Pebble API".

> - Base URL: The consistent foundation of all API endpoints (for example, `http://localhost:4000/v1/`).

Reviewer: I think we can leave out the description "The consistent foundation of all API endpoints" and directly say what the base URL is likely to be in Pebble, especially because most endpoints use the socket. How about something like this:

```suggestion
- Base URL: If you're accessing the API over the Unix socket, the base URL is not applicable. If you're accessing the API over TCP, the base URL is `http://localhost:4000` by default (you can configure the TCP address using `pebble run --http`.
```

Reviewer 2: Agreed on the text change. However, "the base URL" *path* part is applicable, just not the scheme and hostname. Perhaps "the scheme, hostname, and port are not applicable; only the `/v1/` path is important."

> - Path: Specifies the resource or collection being accessed (for example, `/changes`, `/services`). Paths are hierarchical and use forward slashes as separators.

Reviewer: I think we should always include `/v1` in paths, since our concept of base URL is a bit different from normal

- Query parameters: Optional key-value pairs appended to the URL after a question mark (?) and separated by ampersands (&). They provide additional filtering or control over the request (for example, `/services?name=svc1`).

Example:

```bash
http://localhost:4000/v1/services?name=svc1

> ```

Reviewer: I feel we can leave this out. It only works over the socket anyway (if I understood correctly). I'm happy to leave out full example URLs because we have the examples sections elsewhere

### Request

Key components of a request include:

> - HTTP Method: The desired action (for example, `GET` for retrieving data, `POST` for creating data, `PUT` for updating data, and `DELETE` for deleting data).

Reviewer: We could drop PUT and DELETE

> - Headers: Metadata about the request, such as content type, basic auth, and accepted response formats (for example, `Authorization: Basic <credentials>`, `Content-Type: application/json`).

Reviewer: I guess `Authorization` will be needed in the future :smile:, but I think for now we don't need to mention it, since we don't support that header

Reviewer 2: Agreed. I think we can just change this to something like:

> Headers: the only header necessary for Pebble requests the content type. This should be `Content-Type: application/json`.

> - Body: For requests that send data to the server (`POST`, `PUT`), the body contains the data in JSON format.

Reviewer: ```suggestion
- Body: For requests that send data to the server using `POST`, the body contains the data in JSON format.
```

### Response

After processing a request, Pebble sends back a response. Key components of a response are:

- HTTP status code: the result of the request.

> - Headers: metadata about the response, such as content type and caching information.

Reviewer: Does Pebble include caching info?

Reviewer 2: No, again, the only header Pebble sends back is `Content-Type: application/json`.

- Body: data returned by Pebble server in JSON format.

### JSON format

> Pebble API uses JSON for data exchange. Both request and response bodies are formatted as JSON objects or arrays.

Reviewer: Very minor nit, to remove risk of misinterpretation

```suggestion
Pebble API uses JSON for data exchange. Request and response bodies are formatted as JSON objects or arrays.
```

> Example JSON Response:

Reviewer: Shall we say which endpoint this response comes from?

```json
{
  "type": "sync",
  "status-code": 200,
  "status": "OK",
  "result": {
    "name": "svc1",
    "startup": "disabled",
    "current": "stopped"
  }
}
```

### Request types

Pebble API supports both synchronous and asynchronous requests.

- Synchronous requests: The client sends a request and waits for the server to return a response before continuing.
- Asynchronous requests: For long-running operations, Pebble API offers asynchronous processing. The client sends a request and gets an acknowledgment immediately, but the actual processing is still running on the server side. The client can poll for the result. Specific endpoints supporting asynchronous operations are documented below with `type` as `async`.

> ## Using cURL

Reviewer: I think "cURL" is styled as "curl" now, although I also thought it was supposed to be "cURL" at first...

To access API endpoints over the Unix socket, use the `--unix-socket` option of `curl`. For example:

* TODO

    ```
    curl ...
    ```

* TODO

    ```
    curl ...
    ```

## Using the Go client

To use the Go client to access API endpoints over the Unix socket, first create a client using `New`, and then call the methods on the returned Client to interact with the API. Example:

> ```go

Reviewer 2: This is much nicer. However, let's just un-indent it to the top level to avoid the extra indentation.

	if err != nil {
		log.Fatal(err)
	}
	_, err = pebble.WaitChange(changeID, &client.WaitChangeOptions{})
	if err != nil {
		log.Fatal(err)
	}
}
```

We try to never change the underlying HTTP API in a backwards-incompatible way. However, in rare cases we may change the Go client in a backwards-incompatible way.

For more information, see the [Go client documentation](https://pkg.go.dev/github.com/canonical/pebble/client).

## Using the Python client

The Ops library for writing and testing Juju charms includes a [Python client for the Pebble API](https://ops.readthedocs.io/en/latest/reference/pebble.html). You can use the Python client to access API endpoints over the Unix socket. For example:

```python
import ops
client = ops.pebble.Client('/path/to/.pebble.socket')
client.start_services(['svc1', 'svc2'])
```

For more information, see:

- [Ops library documentation for the Python client](https://ops.readthedocs.io/en/latest/reference/pebble.html#ops.pebble.Client)
- [Source code of the Python client](https://github.com/canonical/operator/blob/main/ops/pebble.py)
- [Pebble in the context of Juju charms](https://juju.is/docs/sdk/interact-with-pebble)

> For more examples, see "How to use the Pebble API". <!-- [David] Link to the how-to guide -->

Reviewer: Will we have Python examples in the how-to doc? Maybe we could put this link at the end of the curl section and the Go section instead

Author: For completeness, let's include Python examples in the how-to doc.

Here I didn't mean "for more Python examples (since this is the end of the Python section), see ...", but rather:

```
## Accessing the API
### curl
### go
### python
### more examples - link to how-to
```

But since we removed the `Accessing the API` subsection because of previous comments, now the link is in an awkward position. If we put it under curl, it means "for more curl examples".

How about:

```
## Accessing Pebble API
Here are some methods to access the API. For more examples, see ...
### curl
### go
### python
```

## Data formats

Some API parameters use specific formats as defined below.

### Duration

The format of the timeout duration string is a sequence of decimal numbers, each with optional fraction and a unit suffix, such as "300ms", "-1.5h" or "2h45m".

Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".

### Time

A timestamp as a quoted string in the [RFC 3339](https://datatracker.ietf.org/doc/html/rfc3339) format with sub-second precision. Here are some examples:

- "2006-01-02T15:04:05Z07:00"
- "2006-01-02T15:04:05.999999999Z07:00"
- "2024-12-18T10:45:29.919907089+08:00"

## Errors

API endpoints may return errors as 4xx [status codes](https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml) (client errors) and 5xx status codes (server errors).

Errors are always returned in JSON format with the following structure:

```json
{
  "type": "error",
  "status-code": 500,
  "status": "Internal Server Error",
  "result": {
    "message": "error message",
    "kind": "error kind",
    "value": {},
  },
}
```

Possible values for `status-code`:

> - 400: Bad request. For example, a badly-formatted parameter, like when starting a service, the service name is not provided.
> - 401: Unauthorized.
> - 404: Not found. For example, a change with the specified ID does not exist.
> - 500: Internal server error, for example, the Pebble database is corrupt. If these errors keep happening, please [open an issue](https://github.com/canonical/pebble/issues/new).

Reviewer: A couple of suggestions for structuring each list item the same way:

```suggestion
- 400: Bad request. For example, if a parameter is missing or badly-formatted, such as trying to start a service without providing a the service name.
- 401: Unauthorized.
- 404: Not found. For example, if a change with the specified ID does not exist.
- 500: Internal server error. For example, if the Pebble database is corrupt. If these errors keep happening, please [open an issue](https://github.com/canonical/pebble/issues/new).
```

# API

> <link rel="stylesheet" type="text/css" href="https://unpkg.com/swagger-ui-dist@5.11.0/swagger-ui.css" ></link>

Reviewer 2: This is probably fine for now. But just for reference, the LXD docs import this from the repo into `_static` at build time: https://github.com/canonical/lxd/blob/510c0117cb20591f92126e0b8d3bb583118cd93e/doc/custom_conf.py#L291

# API

Pebble exposes an API to allow remote clients to interact with the daemon. The API can start and stop services, add configuration layers to the plan, and so on.

> ## Accessing the API

Reviewer 2: On reflection, maybe we can just drop this section heading -- this whole doc is a reference for "accessing the API".

However, we do need a section something like "structure of the API" where we mention the different parts of the API in general terms: URL with path or query params, request, response, JSON format, request types (sync, async).

The API uses HTTP over a Unix socket, with access to the API controlled by user ID. If `pebble run` is started with the `--http <address>` option, Pebble exposes a limited set of open-access API endpoints using the given TCP address. See {ref}`api-access-levels`.

> See below for some examples of how to use the API. For more examples, see "How to use the Pebble API". <!-- [David] Link to the how-to guide -->

Reviewer 2: I think we should drop this paragraph (below is also the API reference), and just add a link to the how-to guide under examples when that's ready.

There's a Go client for the API. See the [Go client documentation](https://pkg.go.dev/github.com/canonical/pebble/client) and the examples below.

> We try to never change the underlying HTTP API in a backwards-incompatible way. However, in rare cases we may change the Go client in a backwards-incompatible way.

Reviewer 2: This paragraph is kind of stuck in the middle of mentioning the Go client and the Python client. How about we move it up.

In fact, instead of having these paragraphs here, why don't we make the structure like this:

```
# Accessing the API

## Using curl 

Introductory paragraph and a couple of examples (one GET, one POST).

## Using the Go client

Introductory paragraph and one example (one is probably enough here, just the one from https://pkg.go.dev/github.com/canonical/pebble/client).

## Using the Python client

Paragraph mentioning and linking to the docs (can we link to https://ops.readthedocs.io/en/latest/pebble.html to be more specific?). Maybe a 3-line example like:

     import ops
     client = ops.pebble.Client('/path/to/.pebble.socket')
     client.start_services(['svc1', 'svc2'])
```

There's also a Python client for the API. The Python client is part of the `ops` library used by Juju charms. See the [Ops documentation](https://ops.readthedocs.io/en/latest/).

### Go client examples

TODO: go client code snippets.

### cURL examples

TODO: curl examples.

> ## API access levels

Reviewer 2: For now, can we just drop this section (including "Identities") and have this content at `/explanation/api-and-clients`. I think we should merge the two, but not needed in this PR.

API endpoints fall into one of three access levels, from least restricted to most restricted. See {ref}`api-access-levels`.

### Identities

Pebble has the concept of named "identities", which lets administrators manage users' access to the API. See {ref}`controlling-api-access-using-identities`.

> ## Common parameters

Reviewer 2: How about we call this section "Data formats" or similar? Then have an introductory paragraph like:

> Some API parameters use specific formats as defined below.

We'll also want to describe the time format used in some fields (like `current-since`). RFC3339 with an example.

> ### timeout

Reviewer 2: The parameter field name is "timeout", but what I'd like to document here is the "duration" format. So let's change this to "duration", and then in the API endpoint we say "`timeout` is a [duration](link)".

> The format of the timeout is a decimal number with an optional unit suffix (e.g., "300ms", "1.5s", "2h45m").

Reviewer 2: Can we basically copy this from the Go docs (to mention that it can be a sequence of number+unit.

> A duration string is a possibly signed sequence of decimal numbers, each with optional fraction and a unit suffix, such as "300ms", "-1.5h" or "2h45m". Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".

Author: OK.

FYI, I removed the "possibly signed" part because it doesn't make sense for the timeout (or any parameter that is a duration in Pebble) to be negative. I hope this makes sense.

Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".

> ## Common errors

Reviewer 2: Looking over the rendered docs again, I think we should actually move all error codes to this section, and move them all out of each endpoint. They don't add value there.

So I think this section should be just "Errors" and have an introductory paragraph like so:

> API endpoints may return errors as 4xx status codes (client errors) and 5xx status codes (server errors). Errors are always returned in JSON format with the following structure:
>
>     { ... }

And then just a bulleted list or table of error codes:

* 400: bad request, for example, a badly-formatted parameter
* 404: not found, for example, a change with the specified ID does not exist
* ...
* 500: internal server error, for example, the Pebble database is corrupt; if these errors keep happening, please [open an issue](link)
